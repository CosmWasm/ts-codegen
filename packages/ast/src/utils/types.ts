import * as t from '@babel/types';
import { camel, pascal } from 'case';
import { propertySignature } from './babel';
import { RenderContext } from '../context';
import { JSONSchema } from '@cosmology/ts-codegen-types';

export function getResponseType(
  context: RenderContext,
  underscoreName: string
) {
  const methodName = camel(underscoreName);
  return pascal(
    context.contract?.responses?.[underscoreName]?.title
    ??
    // after v1.1 is adopted, we can deprecate this and require the above response
    `${methodName}Response`
  );
};

const getTypeStrFromRef = ($ref: string): string => {
  if ($ref?.startsWith('#/definitions/')) {
    return $ref.replace('#/definitions/', '');
  }
  throw new Error('what is $ref: ' + $ref);
}

export const getTypeFromRef = ($ref: string) => {
  return t.tsTypeReference(t.identifier(getTypeStrFromRef($ref)))
}

const getArrayTypeFromRef = ($ref: string) => {
  return t.tsArrayType(
    getTypeFromRef($ref)
  );
}


export const getTypeOrRef = (obj: JSONSchema): t.TSType => {
  if (obj.$ref) {
    return getTypeFromRef(obj.$ref);
  } else if (typeof obj.type === 'string') {
    return getType(obj.type);
  } else if (Array.isArray(obj.type)) {
    // Handle array of types by creating a union type
    console.log(obj.type);
    const types = obj.type.map(t => getType(t));
    return t.tsUnionType(types);
  } else {
    throw new Error('Type definition not supported or missing: ' + JSON.stringify(obj));
  }
};

const getArrayTypeFromItems = (
  context: RenderContext,
  items: JSONSchema | JSONSchema[]
): t.TSArrayType => {
  if (Array.isArray(items)) {
    // if (context.options.types.itemsUseTuples === true) {
    // Using the modified method with tsTupleType
    // return t.tsArrayType(
    // t.tsTupleType(items.map(item => getTypeOrRef(item)))
    // );
    // } else {
    // Using the original method with nested arrays
    if (items.length > 0 && items[0].type === 'array') {
      return getArrayTypeFromItems(context, items[0]);
    }
    return items.length > 0 ? t.tsArrayType(t.tsArrayType(getTypeOrRef(items[0]))) : t.tsArrayType(t.tsArrayType(t.tsAnyKeyword()));
    // }
  } else {
    const detect = detectType(items.type);

    if (detect.type === 'array') {
      if (Array.isArray(items.items)) {
        return t.tsArrayType(
          t.tsArrayType(
            getTypeOrRef(items.items[0])
          )
        );
      } else if (items.items) {
        return t.tsArrayType(
          getArrayTypeFromItems(context, items.items)
        );
      } else {
        // Handle the case where items.items is undefined for array types
        return t.tsArrayType(t.tsAnyKeyword())
      }
    }

    return t.tsArrayType(
      getType(detect.type)
    );
  }
};

export const detectType = (type: string | string[]) => {
  let optional = false;
  let theType = '';
  if (Array.isArray(type)) {
    if (type.length !== 2) {
      throw new Error('[getType(array length)] case not handled by transpiler. contact maintainers.')
    }
    const [nullableType, nullType] = type;
    if (nullType !== 'null') {
      throw new Error('[getType(null)] case not handled by transpiler. contact maintainers.')
    }
    theType = nullableType;
    optional = true;
  } else {
    theType = type;
  }

  return {
    type: theType,
    optional
  };
}

export const getTypeInfo = (
  context: RenderContext,
  info: JSONSchema
) => {
  let type = undefined;
  let optional = undefined;

  // Handle nullable types using 'anyOf' schema
  if (Array.isArray(info.anyOf)) {
    if (info.anyOf.length !== 2) {
      throw new Error('case not handled by transpiler. contact maintainers.');
    }
    const [nullableType, nullType] = info.anyOf;
    if (nullType?.type !== 'null') {
      throw new Error('[nullableType.type]: case not handled by transpiler. contact maintainers.');
    }
    if (!nullableType?.$ref) {
      if (nullableType.title) {
        type = t.tsTypeReference(t.identifier(nullableType.title));
      } else {
        throw new Error('[nullableType.title] case not handled by transpiler. contact maintainers.');
      }
    } else {
      type = getTypeFromRef(nullableType.$ref);
    }
    optional = true;
  } else if (typeof info.type === 'string') {
    console.log('is string')
    // Process type when it's explicitly defined
    if (info.type === 'array') {
      if (typeof info.items === 'object' && !Array.isArray(info.items)) {
        if (info.items.$ref) {
          type = getArrayTypeFromRef(info.items.$ref);
        } else if (info.items.title) {
          type = t.tsArrayType(t.tsTypeReference(t.identifier(info.items.title)));
        } else if (info.items.type) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
        // type = getArrayTypeFromItems(context, info.items);
      } else {
        if (Array.isArray(info.items)) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
      }
    } else {
      const detect = detectType(info.type);
      type = getType(detect.type);
      optional = detect.optional;
    }
  } else if (Array.isArray(info.type)) {
    console.log('is arr', info.type)
    // Handle multiple types, typically nullable types
    if (info.type.length !== 2) {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }
    const [nullableType, nullType] = info.type;
    if (nullType !== 'null') {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }

    if (nullableType === 'array' && typeof info.items === 'object' && !Array.isArray(info.items)) {
      if (info.items.type) {
        const detect = detectType(info.items.type);
        if (detect.type === 'array') {
          console.log('items type')
          type = t.tsArrayType(getArrayTypeFromItems(context, info.items));
        } else {
          type = t.tsArrayType(getType(detect.type));
        }
        optional = detect.optional;
      } else if (info.items.$ref) {
        type = getArrayTypeFromRef(info.items.$ref);
      } else if (info.items.type) {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
      }
    } else {
      const detect = detectType(nullableType);
      optional = detect.optional;
      if (detect.type === 'array') {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        type = getType(detect.type);
      }
    }

    optional = true;  // Ensure optional is set for nullable types
  }
  console.log({ type })

  return {
    type,
    optional
  };
};

export const getType = (type: string) => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'boolean':
      return t.tSBooleanKeyword();
    case 'integer':
      return t.tsNumberKeyword();
    case 'null':
      return t.tsNullKeyword();
    case 'number':
      return t.tsNumberKeyword();
    default:
      throw new Error('contact maintainers [unknown type]: ' + type);
  }
}

export const getPropertyType = (
  context: RenderContext,
  schema: JSONSchema,
  prop: string
) => {
  const props = schema.properties ?? {};
  let info = props[prop];

  let type = null;
  let optional = !schema.required?.includes(prop);

  if (info.allOf && info.allOf.length === 1) {
    info = info.allOf[0];
  }

  if (typeof info.$ref === 'string') {
    type = getTypeFromRef(info.$ref)
  }

  const typeInfo = getTypeInfo(context, info);
  if (typeof typeInfo.optional !== 'undefined') {
    optional = typeInfo.optional;
  }
  if (typeof typeInfo.type !== 'undefined') {
    console.log(JSON.stringify(typeInfo, null, 2))
    type = typeInfo.type;
  }

  if (!type) {
    throw new Error('cannot find type for ' + JSON.stringify(info))
  }

  if (schema.required?.includes(prop)) {
    optional = false;
  }

  return { type, optional };
};


export function getPropertySignatureFromProp(
  context: RenderContext,
  jsonschema: JSONSchema,
  prop: string,
  camelize: boolean
) {
  if (jsonschema.properties[prop].type === 'object') {
    if (jsonschema.properties[prop].title) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(t.identifier(jsonschema.properties[prop].title))
        )
      );
    } else {
      throw new Error('getPropertySignatureFromProp() contact maintainer');
    }
  }

  if (Array.isArray(jsonschema.properties[prop].allOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].allOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    // @ts-ignore:next-line
    const uniqUnionTypes = [...new Set(unionTypes)];

    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }
  } else if (Array.isArray(jsonschema.properties[prop].oneOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].oneOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    const uniqUnionTypes = [...new Set(unionTypes)];
    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }

  }

  try {
    getPropertyType(context, jsonschema, prop);
  } catch (e) {
    console.log(e);
    console.log(JSON.stringify(jsonschema, null, 2), prop);
  }

  const { type, optional } = getPropertyType(context, jsonschema, prop);
  return propertySignature(
    camelize ? camel(prop) : prop,
    t.tsTypeAnnotation(
      type
    ),
    optional
  );
}

export const getParamsTypeAnnotation = (
  context: RenderContext,
  jsonschema: any,
  camelize: boolean = true
): t.TSTypeAnnotation => {
  const keys = Object.keys(jsonschema.properties ?? {});

  if (!keys.length && jsonschema.$ref) {
    return t.tsTypeAnnotation(getTypeFromRef(jsonschema.$ref))
  }

  if (!keys.length) return undefined;

  const typedParams = keys.map(prop => getPropertySignatureFromProp(
    context,
    jsonschema,
    prop,
    camelize
  ));

  return t.tsTypeAnnotation(
    t.tsTypeLiteral(
      // @ts-ignore:next-line
      [
        ...typedParams
      ]
    )
  )
}

export const createTypedObjectParams = (
  context: RenderContext,
  jsonschema: JSONSchema,
  camelize: boolean = true
): (t.Identifier | t.Pattern | t.RestElement) => {

  const keys = Object.keys(jsonschema.properties ?? {});
  if (!keys.length) {
    // is there a ref?
    if (jsonschema.$ref) {
      const obj = context.refLookup(jsonschema.$ref);
      // If there is a oneOf, then we need to create a type for it
      if (obj?.oneOf) {
        // the actual type of the ref
        const refType = jsonschema.$ref.split('/').pop();
        const refName = camel(refType);
        const id = t.identifier(refName);
        id.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(refType)));
        return id
      } else if (obj) {
        return createTypedObjectParams(
          context,
          obj,
          camelize
        );
      }
    }

    // no results...
    return;
  }

  const params = keys.map(prop => {
    return t.objectProperty(
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      false,
      true
    );
  });

  const obj = t.objectPattern(
    [
      ...params
    ]
  );

  obj.typeAnnotation = getParamsTypeAnnotation(
    context,
    jsonschema,
    camelize
  );

  return obj;
};