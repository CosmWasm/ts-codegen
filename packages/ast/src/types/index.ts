export * from './context';
export * from './plugins';