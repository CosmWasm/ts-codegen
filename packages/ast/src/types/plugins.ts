export interface ReactQueryOptions {
  enabled?: boolean;
  optionalClient?: boolean;
  version?: 'v3' | 'v4';
  mutations?: boolean;
  camelize?: boolean;
  queryKeys?: boolean
  queryFactory?: boolean
}

export interface TSClientOptions {
  enabled?: boolean;
  execExtendsQuery?: boolean;
  noImplicitOverride?: boolean;
}
export interface MessageComposerOptions {
  enabled?: boolean;
}
export interface MessageBuilderOptions {
  enabled?: boolean;
}
export interface RecoilOptions {
  enabled?: boolean;
}
export interface TSTypesOptions {
  enabled?: boolean;
  itemsUseTuples?: boolean;
  // deprecated
  aliasExecuteMsg?: boolean;
  aliasEntryPoints?: boolean;
}

export interface UseContractsOptions {
  enabled?: boolean;
};
