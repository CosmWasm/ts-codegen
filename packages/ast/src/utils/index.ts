export * from './babel';
export { OPTIONAL_FUNDS_PARAM } from './constants';
export { FIXED_EXECUTE_PARAMS } from './constants';
export { PROVIDER_TYPES } from './constants';
export * from './ref';
export * from './types';
