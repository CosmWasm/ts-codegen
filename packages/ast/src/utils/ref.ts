import { JSONSchema } from "@cosmology/ts-codegen-types";

export const refLookup = ($ref: string, schema: JSONSchema): JSONSchema => {
  const refName = $ref.replace('#/definitions/', '')
  return schema.definitions?.[refName];
}