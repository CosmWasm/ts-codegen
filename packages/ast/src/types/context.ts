import { IDLObject, JSONSchema } from "@cosmwasm/ts-codegen-types";
import { MessageBuilderOptions, MessageComposerOptions, ReactQueryOptions, RecoilOptions, TSClientOptions, TSTypesOptions, UseContractsOptions } from "./plugins";

export interface ContractInfo {
  schemas: JSONSchema[];
  responses?: Record<string, JSONSchema>;
  idlObject?: Partial<IDLObject>;
};
export interface RenderOptions {
  enabled?: boolean;
  types?: TSTypesOptions;
  recoil?: RecoilOptions;
  messageComposer?: MessageComposerOptions;
  messageBuilder?: MessageBuilderOptions;
  client?: TSClientOptions;
  reactQuery?: ReactQueryOptions;
  useContractsHook?: UseContractsOptions;
}

export interface ProviderInfo {
  classname: string,
  filename: string,
  basename: string,
}
