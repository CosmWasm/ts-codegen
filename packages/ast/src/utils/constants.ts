import { identifier } from './babel';
import * as t from '@babel/types';

export const OPTIONAL_FUNDS_PARAM = identifier(
  '_funds',
  t.tsTypeAnnotation(t.tsArrayType(t.tsTypeReference(t.identifier('Coin')))),
  true
);
export const OPTIONAL_FEE_PARAM = identifier(
  'fee',
  t.tsTypeAnnotation(
    t.tsUnionType([
      t.tsNumberKeyword(),
      t.tsTypeReference(t.identifier('StdFee')),
      t.tsLiteralType(t.stringLiteral('auto'))
    ])
  ),
  true
);
export const OPTIONAL_MEMO_PARAM = identifier(
  'memo',
  t.tsTypeAnnotation(t.tsStringKeyword()),
  true
);

export const FIXED_EXECUTE_PARAMS = [
  OPTIONAL_FEE_PARAM,
  OPTIONAL_MEMO_PARAM,
  OPTIONAL_FUNDS_PARAM
];

export const PROVIDER_TYPES = {
  SIGNING_CLIENT_TYPE: "client",
  QUERY_CLIENT_TYPE: "queryClient",
  MESSAGE_COMPOSER_TYPE : 'message-composer',
  PROVIDER_TYPE : 'provider'
};
