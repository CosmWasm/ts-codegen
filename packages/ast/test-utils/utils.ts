import generate from '@babel/generator';
import { sync as glob } from 'glob';
import { readFileSync } from 'fs';
import { join } from 'path';
import { JSONSchema } from '../src/types';
import { RenderContext, RenderOptions } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
}

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
}

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options)
};

interface GlobContract {
  name: `/${string}.json`;
  content: JSONSchema;
}

export const globContracts = (p: string): GlobContract[] => {
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(__dirname, '/../../../__fixtures__/', p, '/*.json')).map(file => {
    return { name: file.split('__fixtures__')[1], content: JSON.parse(readFileSync(file, 'utf-8')) }
  });
  return contracts;
};
