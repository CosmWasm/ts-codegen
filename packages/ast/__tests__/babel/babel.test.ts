import * as t from '@babel/types';

import { importStmt } from '../../src/utils/babel';
import {
  arrowFunctionExpression,
  bindMethod,
  classDeclaration,
  classProperty,
  promiseTypeAnnotation,
  typedIdentifier
} from '../../src/utils/babel';
import { expectCode } from '../../test-utils';

it('top import', async () => {
  expectCode(
    importStmt(
      ['CosmWasmClient', 'ExecuteResult', 'SigningCosmWasmClient'],
      '@cosmjs/cosmwasm-stargate'
    )
  );
});

it('interfaces', async () => {
  expectCode(
    t.program([
      t.exportNamedDeclaration(
        t.tsInterfaceDeclaration(
          t.identifier('SG721ReadOnlyInstance'),
          null,
          [],
          t.tSInterfaceBody([
            t.tSPropertySignature(
              t.identifier('contractAddress'),
              t.tsTypeAnnotation(t.tsStringKeyword())
            ),
            // methods

            t.tSPropertySignature(
              t.identifier('tokens'),
              t.tsTypeAnnotation(
                t.tsFunctionType(
                  null,
                  [
                    typedIdentifier(
                      'owner',
                      t.tsTypeAnnotation(t.tsStringKeyword())
                    ),
                    typedIdentifier(
                      'startAfter',
                      t.tsTypeAnnotation(t.tsStringKeyword()),
                      true
                    ),
                    typedIdentifier(
                      'limit',
                      t.tsTypeAnnotation(t.tsStringKeyword()),
                      true
                    )
                  ],
                  promiseTypeAnnotation('TokensResponse')
                )
              )
            )
          ])
        )
      ),

      // extends

      t.exportNamedDeclaration(
        t.tsInterfaceDeclaration(
          t.identifier('SG721Instance'),
          null,
          [
            t.tSExpressionWithTypeArguments(
              t.identifier('SG721ReadOnlyInstance')
            )
          ],
          t.tSInterfaceBody([
            // contract address

            t.tSPropertySignature(
              t.identifier('contractAddress'),
              t.tsTypeAnnotation(t.tsStringKeyword())
            ),

            // METHOD
            t.tSPropertySignature(
              t.identifier('mint'),
              t.tsTypeAnnotation(
                t.tsFunctionType(
                  null,
                  [
                    typedIdentifier(
                      'sender',
                      t.tsTypeAnnotation(t.tsStringKeyword())
                    ),
                    typedIdentifier(
                      'anotherProp',
                      t.tsTypeAnnotation(t.tsStringKeyword())
                    ),
                    typedIdentifier(
                      'prop3',
                      t.tsTypeAnnotation(t.tsStringKeyword())
                    )
                  ],
                  promiseTypeAnnotation('ExecuteResult')
                )
              )
            )
          ])
        )
      )
    ])
  );
});

it('readonly classes', async () => {
  expectCode(
    t.program([
      t.exportNamedDeclaration(
        classDeclaration(
          'SG721QueryClient',
          [
            // client
            classProperty(
              'client',
              t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier('CosmWasmClient'))
              )
            ),

            // contractAddress
            classProperty(
              'contractAddress',
              t.tsTypeAnnotation(t.tsStringKeyword())
            ),

            // constructor
            t.classMethod(
              'constructor',
              t.identifier('constructor'),
              [
                typedIdentifier(
                  'client',
                  t.tsTypeAnnotation(
                    t.tsTypeReference(t.identifier('CosmWasmClient'))
                  )
                ),
                typedIdentifier(
                  'contractAddress',
                  t.tsTypeAnnotation(t.tsStringKeyword())
                )
              ],
              t.blockStatement([
                // client/contract set
                t.expressionStatement(
                  t.assignmentExpression(
                    '=',
                    t.memberExpression(
                      t.thisExpression(),
                      t.identifier('client')
                    ),
                    t.identifier('client')
                  )
                ),
                t.expressionStatement(
                  t.assignmentExpression(
                    '=',
                    t.memberExpression(
                      t.thisExpression(),
                      t.identifier('contractAddress')
                    ),
                    t.identifier('contractAddress')
                  )
                ),

                // bindings
                bindMethod('approval'),
                bindMethod('otherProp'),
                bindMethod('hello'),
                bindMethod('mintme')
              ])
            ),

            // methods:
            t.classProperty(
              t.identifier('approval'),
              arrowFunctionExpression(
                [
                  // props
                  typedIdentifier(
                    'owner',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  ),
                  //
                  typedIdentifier(
                    'spender',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  )
                ],
                t.blockStatement([
                  t.returnStatement(
                    t.callExpression(
                      t.memberExpression(
                        t.memberExpression(
                          t.thisExpression(),
                          t.identifier('client')
                        ),
                        t.identifier('queryContractSmart')
                      ),
                      [
                        t.memberExpression(
                          t.thisExpression(),
                          t.identifier('contractAddress')
                        ),
                        t.objectExpression([
                          // props
                          t.objectProperty(
                            t.identifier('owner'),
                            t.identifier('owner'),
                            false,
                            true
                          ),

                          t.objectProperty(
                            t.identifier('spender'),
                            t.identifier('spender'),
                            false,
                            true
                          )
                        ])
                      ]
                    )
                  )
                ]),
                t.tsTypeAnnotation(
                  t.tsTypeReference(
                    t.identifier('Promise'),
                    t.tsTypeParameterInstantiation([
                      t.tSTypeReference(t.identifier('ApprovalResponse'))
                    ])
                  )
                ),
                true
              )
            )
          ],
          [
            t.tSExpressionWithTypeArguments(
              t.identifier('SG721ReadOnlyInstance')
            )
          ]
        )
      )
    ])
  );
});

it('mutation classes', async () => {
  expectCode(
    t.program([
      t.exportNamedDeclaration(
        classDeclaration(
          'SG721Client',
          [
            // client
            classProperty(
              'client',
              t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier('SigningCosmWasmClient'))
              )
            ),

            // contractAddress
            classProperty(
              'contractAddress',
              t.tsTypeAnnotation(t.tsStringKeyword())
            ),

            // constructor
            t.classMethod(
              'constructor',
              t.identifier('constructor'),
              [
                typedIdentifier(
                  'client',
                  t.tsTypeAnnotation(
                    t.tsTypeReference(t.identifier('SigningCosmWasmClient'))
                  )
                ),
                typedIdentifier(
                  'contractAddress',
                  t.tsTypeAnnotation(t.tsStringKeyword())
                )
              ],
              t.blockStatement([
                // super()
                t.expressionStatement(
                  t.callExpression(t.super(), [
                    t.identifier('client'),
                    t.identifier('contractAddress')
                  ])
                ),

                // client/contract set
                t.expressionStatement(
                  t.assignmentExpression(
                    '=',
                    t.memberExpression(
                      t.thisExpression(),
                      t.identifier('client')
                    ),
                    t.identifier('client')
                  )
                ),
                t.expressionStatement(
                  t.assignmentExpression(
                    '=',
                    t.memberExpression(
                      t.thisExpression(),
                      t.identifier('contractAddress')
                    ),
                    t.identifier('contractAddress')
                  )
                ),

                // bindings
                bindMethod('approval'),
                bindMethod('otherProp'),
                bindMethod('hello'),
                bindMethod('mintme')
              ])
            ),

            // methods:
            t.classProperty(
              t.identifier('mint'),
              arrowFunctionExpression(
                [
                  // props
                  typedIdentifier(
                    'sender',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  ),
                  typedIdentifier(
                    'tokenId',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  ),
                  typedIdentifier(
                    'owner',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  ),
                  typedIdentifier(
                    'token_uri',
                    t.tsTypeAnnotation(t.tsStringKeyword())
                  )
                ],
                t.blockStatement([
                  t.returnStatement(
                    t.awaitExpression(
                      t.callExpression(
                        t.memberExpression(
                          t.memberExpression(
                            t.thisExpression(),
                            t.identifier('client')
                          ),
                          t.identifier('execute')
                        ),
                        [
                          t.identifier('sender'),
                          t.memberExpression(
                            t.thisExpression(),
                            t.identifier('contractAddress')
                          ),
                          t.objectExpression([
                            t.objectProperty(
                              t.identifier('mint'),
                              t.objectExpression([
                                t.objectProperty(
                                  t.identifier('token_id'),
                                  t.identifier('tokenId')
                                ),
                                t.objectProperty(
                                  t.identifier('owner'),
                                  t.identifier('owner')
                                ),
                                t.objectProperty(
                                  t.identifier('token_uri'),
                                  t.identifier('token_uri')
                                ),
                                t.objectProperty(
                                  t.identifier('expression'),
                                  t.objectExpression([])
                                )
                              ])
                            )
                          ]),
                          t.stringLiteral('auto')
                        ]
                      )
                    )
                  )
                ]),
                // return type
                t.tsTypeAnnotation(
                  t.tsTypeReference(
                    t.identifier('Promise'),
                    t.tsTypeParameterInstantiation([
                      t.tSTypeReference(t.identifier('ExecuteResult'))
                    ])
                  )
                ),
                true
              )
            )
          ],
          [
            t.tSExpressionWithTypeArguments(
              t.identifier('SG721ReadOnlyInstance')
            )
          ],
          t.identifier('SG721QueryClient')
        )
      )
    ])
  );
});

it('object parms', () => {
  const obj = t.objectPattern([
    t.objectProperty(
      t.identifier('includeExpired'),
      t.identifier('includeExpired'),
      false,
      true
    ),
    t.objectProperty(
      t.identifier('tokenId'),
      t.identifier('tokenId'),
      false,
      true
    )
  ]);
  obj.typeAnnotation = t.tsTypeAnnotation(
    t.tsTypeLiteral([
      t.tsPropertySignature(
        t.identifier('includeExpired'),
        t.tsTypeAnnotation(t.tsBooleanKeyword())
      ),
      t.tsPropertySignature(
        t.identifier('tokenId'),
        t.tsTypeAnnotation(t.tsStringKeyword())
      )
    ])
  );
  expectCode(
    t.program([
      t.expressionStatement(
        t.assignmentExpression(
          '=',
          t.identifier('ownerOf'),
          arrowFunctionExpression(
            [obj],
            t.blockStatement(
              [
                // body
              ],
              []
            ),
            t.tsTypeAnnotation(
              t.tsTypeReference(
                t.identifier('Promise'),
                t.tsTypeParameterInstantiation([
                  t.tsTypeReference(t.identifier('OwnerOfResponse'))
                ])
              )
            )
          )
        )
      )
    ])
  );
});
