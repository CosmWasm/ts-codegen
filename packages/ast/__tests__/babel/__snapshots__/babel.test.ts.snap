// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interfaces 1`] = `
"export interface SG721ReadOnlyInstance {
  contractAddress: string;
  tokens: (owner: string, startAfter?: string, limit?: string) => Promise<TokensResponse>;
}
export interface SG721Instance extends SG721ReadOnlyInstance {
  contractAddress: string;
  mint: (sender: string, anotherProp: string, prop3: string) => Promise<ExecuteResult>;
}"
`;

exports[`mutation classes 1`] = `
"export class SG721Client extends SG721QueryClient implements SG721ReadOnlyInstance {
  client: SigningCosmWasmClient;
  contractAddress: string;

  constructor(client: SigningCosmWasmClient, contractAddress: string) {
    super(client, contractAddress);
    this.client = client;
    this.contractAddress = contractAddress;
    this.approval = this.approval.bind(this);
    this.otherProp = this.otherProp.bind(this);
    this.hello = this.hello.bind(this);
    this.mintme = this.mintme.bind(this);
  }

  mint = async (sender: string, tokenId: string, owner: string, token_uri: string): Promise<ExecuteResult> => {
    return await this.client.execute(sender, this.contractAddress, {
      mint: {
        token_id: tokenId,
        owner: owner,
        token_uri: token_uri,
        expression: {}
      }
    }, \\"auto\\");
  };
}"
`;

exports[`object parms 1`] = `
"ownerOf = ({
  includeExpired,
  tokenId
}: {
  includeExpired: boolean;
  tokenId: string;
}): Promise<OwnerOfResponse> => {};"
`;

exports[`readonly classes 1`] = `
"export class SG721QueryClient implements SG721ReadOnlyInstance {
  client: CosmWasmClient;
  contractAddress: string;

  constructor(client: CosmWasmClient, contractAddress: string) {
    this.client = client;
    this.contractAddress = contractAddress;
    this.approval = this.approval.bind(this);
    this.otherProp = this.otherProp.bind(this);
    this.hello = this.hello.bind(this);
    this.mintme = this.mintme.bind(this);
  }

  approval = async (owner: string, spender: string): Promise<ApprovalResponse> => {
    return this.client.queryContractSmart(this.contractAddress, {
      owner,
      spender
    });
  };
}"
`;

exports[`top import 1`] = `"import { CosmWasmClient, ExecuteResult, SigningCosmWasmClient } from \\"@cosmjs/cosmwasm-stargate\\";"`;
